### PR TITLE
feat: user-defined calculated channels with persisted formula/input preferences

### LINKED ISSUE #111: Add Support for User-Defined Calculated Channels
Resolves #113 

## Description

Introduce support for **user-defined calculated channels** created from existing channels within a specific **Channel File**.

Users must be able to define mathematical expressions combining one or more channels and generate a derived channel that behaves like a standard channel.

#### Option A — From Channel File
- When right-clicking on a Channel File, show a context menu option:
  - **Add Calculated Channel**

#### Option B — From Calculated Channels Section
Under each Channel File, always show a section at the bottom of the channel list:
- **Calculated Channels (0)**

Requirements:
- Must always be visible after loading a Channel File.
- Must appear at the very bottom of the channel list.
- Must be collapsed by default.
- If no calculated channels exist, it must display `Calculated Channels (0)` with no child items.
- This section must include a **"+"** button aligned with the section header.
  - Tooltip: **Add Calculated Channel**
  - Clicking this button opens the same configuration dialog as the context menu option.
- Right-clicking the **Calculated Channels** section must show only:
  - **Add Calculated Channel**
- Both entry points must open the same dialog.

#### Default number of inputs
- Default number of channel inputs must always be **2**.
- If the user changes it (e.g., to 3), that change must **not persist**.
- Creating a new calculated channel must always reset the default input count back to **2**.

#### Channel Title
- Text field defining the new channel name.

#### Number of Channel Inputs
- Integer selector between **0 and 10**.
- Default value: **2**.
- Not persisted between creations.
- Behavior:
  - If set to **0**, the formula represents a constant (e.g., `4`).

#### Channel Selectors
- If input count > 0:
  - Labels auto-generated: **A, B, C…** (up to 10).
  - Each label has a searchable dropdown listing channels from the same Channel File.
  - Type-to-filter behavior.

#### Formula Field
- Symbolic expression referencing **A, B, C…**
- Allowed operators:
  - `+`
  - `-`
  - `*`
  - `/`
  - Parentheses `()`
- Must support unary minus for sign inversion.

Valid examples:
- `A + B`
- `A / (B + C)`
- `-A`
- `-(A + B)`
- `(A - B) * C`
- `4` (constant mode)

#### Units
- Optional text field.
- Default: empty.

#### Description
- Optional text field.
- Displayed as tooltip when hovering over the calculated channel.

### 4) Offset and Gain Awareness
When evaluating calculated channels:
- If any source channel has an applied offset or gain, use transformed values (after gain/offset).
- Calculated channel values must reflect exactly what the user sees in plots.

### 5) Formula Validation
Include a **Validate Formula** button that performs:

#### 5.1 Channel assignment check
- No required selector may be empty.

#### 5.2 Data integrity check
- Scan all values across selected channels.
- Reject if any `NaN` is found.

#### 5.3 Syntax validation
Reject:
- Incomplete expressions (e.g., `A +`)
- Invalid operator sequences (`+ *`)
- Unbalanced parentheses
- Invalid tokens
- Unsupported operators

Must support unary minus correctly.

#### 5.4 Division-by-zero check
- If division exists, evaluate denominator across the entire time series.
- If denominator equals zero at any timestamp, reject validation and show an appropriate error.

Example:
- `A / (B + C)` → evaluate `(B + C)` fully; if any zero is detected, validation fails.

### 6) Channel Creation and Storage
After successful validation:
- Create the calculated channel.
- Add it under:
  - **Calculated Channels (N)**
- `N` updates dynamically.
- Section remains at bottom of channel list.

Calculated channels must:
- Behave like standard channels.
- Be addable to plots.
- Support zoom, styling, export, etc.
- Be searchable in the search bar from that channel file.
- Use incremental numbers from the last original channel.
  - Example: if original channels end at `10`, new calculated channels are `11`, `12`, …

### 7) Deletion Behavior
If the parent Channel File is deleted:
- All its calculated channels must be deleted.
- Any plot references must be removed automatically.

### 8) Editing Calculated Channels
Right-click on a calculated channel:
- Provide **Edit** option.
- Open same dialog with all fields pre-filled.

Allow modification of:
- Title
- Inputs
- Formula
- Units
- Description

Require revalidation before saving changes.

## Acceptance Criteria
- Default input count is always `2`.
- Input count does not persist across new creations.
- Calculated Channels section always appears (even with 0 items).
- Section is collapsed by default.
- `+` button opens creation dialog.
- Context menu from section supports creation.
- Searchable dropdown works correctly.
- Unary minus (`-A`) is supported.
- Gain and offset are respected in calculations.
- Syntax validation catches malformed expressions.
- Division-by-zero is validated across time.
- Calculated channels behave like normal channels.
- Deleting Channel File removes calculated channels.
- Editing works correctly and revalidates.
- No regression in existing channel handling.

---

## Implementation Update (2026-02-28)

Implemented and validated in branch `feat/add_calculated_channels` (PR: https://github.com/ESPlotter/ESPlotter/pull/113).

### Delivered behavior
- Added user-defined calculated channels with create/edit/delete flows in sidebar.
- Added formula parser/evaluator support including unary minus and validation checks.
- Added circular dependency prevention (self-reference and indirect cycles).
- Added duplicate title validation against original and calculated channels.
- Added rename propagation to chart/page contexts.
- Added search empty-state refinement for filtered calculated channels.
- Added per-input-count persisted formula memory (`0..10`).
- Added persisted last input count (`0..10`) so create dialog preloads previous value.
- Formula field defaults to blank when empty and selects all text on focus/click.

### Testing status
- Unit suites (main/main-infrastructure/preload/renderer): passing locally.
- Renderer component suite for AppSidebar: passing locally.
- Formula utility unit tests expanded with comprehensive edge cases: passing locally.

### Scope note
- During implementation, product scope was updated to persist both:
  - last formula by input count
  - last selected input count
  across app restarts.

---

## Implementation Update (2026-03-02)

Additional implementation/review passes completed in the same branch/PR:

### Additional delivered behavior and hardening
- Fixed channel refresh flow after applying file time-offset, channel gain, and channel offset by using the latest ready-file snapshot before chart refresh.
- Hardened formula evaluator error context when `timeValues` contain `null` by falling back safely to sample index.
- Improved calculated-channel deletion propagation and cleanup:
  - store-level cascading deletion handles dependents
  - delete action now returns all removed IDs
  - renderer cleanup removes all removed IDs from all pages/charts.
- Removed unused `isCalculatedDialogSelected` state from `AppSidebar`.
- Stabilized E2E spinner behavior check without weakening contract:
  - test keeps strict spinner assertion
  - main open flow includes an opt-in E2E delay via `ESPLOTTER_E2E_OPEN_DELAY_MS`
  - spinner test sets that env var explicitly to avoid CI race flakiness.

### Test coverage updates
- Added/updated renderer unit and component regressions around calculated channel behavior and formula parsing edge cases.
- Updated store tests to validate returned removed IDs from cascading deletes.
- Stabilized spinner E2E coverage while preserving assertion strength.

### Validation status (latest)
- `npm run test:unit`
- `npm run test:components`
- `npm run test:e2e -- --grep-invert "example\.out|Open \.out files"`
- `npm run test:e2e -- e2e/open-file.spec.ts`

All listed runs passed in local validation after the fixes above.

### PR DESCRIPTION
**Pull request by jarosat**
Resolves #111 

## Summary
- Implements user-defined calculated channels end-to-end in sidebar/dialog flows.
- Adds validation and UX improvements (duplicate names, circular dependency prevention, search empty states, selector keyboard UX, formula help/indicators).
- Propagates calculated-channel rename/remove updates across pages/charts/series contexts.
- Persists calculated-channel preferences in user settings:
  - last formula by input count (`0..10`)
  - last selected input count (`0..10`) used to prefill next create dialog.
- Extends IPC/preload/main/renderer contracts and observers for these preferences.
- Expands formula utility coverage with parser/evaluator edge cases.

## Additional fixes after review/QA
- Fixed stale refresh path after file time offset / channel gain / channel offset changes by refreshing with latest ready-file snapshot.
- Fixed formula error time annotation when `timeValues` has `null` values.
- Improved cascading delete behavior in channel files store and propagated cleanup to all pages/charts:
  - delete action now returns all removed IDs
  - caller removes every removed channel key from all pages.
- Removed unused `isCalculatedDialogSelected` state in `AppSidebar`.
- Hardened spinner-related E2E reliability without weakening checks:
  - spinner test still asserts visibility (`spinnerWasVisible === true`)
  - `waitForFileSpinner` only swallows timeout errors and rethrows others
  - added opt-in E2E open delay (`ESPLOTTER_E2E_OPEN_DELAY_MS`) used by spinner test to avoid CI race.

## Testing
- `npm run test:unit`
- `npm run test:components`
- `npm run test:e2e -- --grep-invert "example\.out|Open \.out files"`
- `npm run test:e2e -- e2e/open-file.spec.ts`
- Focused suites also run during review iterations:
  - `npm run test:unit:renderer -- src/test/renderer/channel-file/store/ChannelFilesStore.spec.ts`
  - `npm run test:components -- src/test/renderer/app-sidebar/components/AppSidebar/AppSidebar.spec.tsx`

## Notes
- `.out` flow remains intentionally excluded from the broad E2E run in this validation set.
- CodeRabbit feedback addressed with follow-up commits, including the final spinner CI stabilization.

## Summary by CodeRabbit

* **New Features**
  * Floating, draggable/resizable Calculated Channel dialog with per-input controls, keyboard navigation, validation, icons, and create/edit/delete/add actions.
  * Persistence of last-used input count and per-input formulas; preferences exposed and synced.
  * SQRT and PI support in formulas with improved validation and time-aware error messages.
  * Calculated-channel create/delete/rename propagate across pages and clean up dependent channels.

* **Style**
  * Optional tooltip arrows.

### CODE REVIEW

**Comment by jarosat** [comment]
Implemented the open CodeRabbit feedback on this PR and pushed the changes in commit `f415cd0`.

### ✅ Applied changes
1. **Legacy preferences safety**
   - Added defensive fallback in `UserPreferences.fromPrimitives` when `general.calculatedChannels` is missing (older persisted data).

2. **Calculated channel delete across all pages**
   - Added a cross-page action in `PlotPagesStore` to remove a single channel key from every page/chart.
   - Updated `AppSidebar` calculated delete flow to use this cross-page removal.

3. **Create/Edit save flow robustness**
   - Made calculated-channel preference persistence non-blocking after a successful channel save.
   - Channel create/edit success now closes dialog and shows success even if preference persistence fails later.

4. **Calculated channel ID monotonicity**
   - Updated `addCalculatedChannel` to compute next number from max existing calculated number/id suffix, preventing reuse after deletions.

5. **Division guard precision**
   - Replaced the denominator check with strict zero detection (`right === 0`) so very small non-zero values are treated as valid.

6. **Repository mock mutation safety in tests**
   - `UserPreferencesRepositoryMock` now clones on `get`, `save`, and `setPreferences` to avoid in-place mutation masking change detection.

### ✅ Regression tests added/updated
- AppSidebar component tests:
  - creation remains successful when preference persistence fails
  - calculated channel delete triggers cross-page removal
- ChannelFilesStore tests:
  - no calculated channel number/id reuse after deletion
- Formula utils tests:
  - tiny non-zero denominator is allowed

### ✅ Verification run locally
- `npm run test:components -- src/test/renderer/app-sidebar/components/AppSidebar/AppSidebar.spec.tsx`
- `npm run test:unit:renderer -- src/test/renderer/channel-file/store/ChannelFilesStore.spec.ts src/test/renderer/channel-file/utils/calculatedChannels.spec.ts`
- `npm run test:unit:main -- src/test/main/user-preferences`
- `npm run test:unit:main-infrastructure -- src/test/main/user-preferences/infrastructure`

All passed.

I did not intentionally skip any currently open high/major CodeRabbit findings from this PR thread.

File: src/test/main/user-preferences/infrastructure/ipc/registerUserPreferencesIpcHandlers.spec.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  beforeEach(() => {
+    registeredHandlers.clear();
+    vi.clearAllMocks();
+    registerUserPreferencesIpcHandlers();
+  });
+
+  it('should register calculated channel last input count handlers', () => {
+    expect(registeredHandlers.has('getCalculatedChannelLastInputCount')).toBe(true);
+    expect(registeredHandlers.has('updateCalculatedChannelLastInputCount')).toBe(true);
+  });
```
[the remark this answers is not in the document]
Comment (jarosat) [reply src/test/main/user-preferences/infrastructure/ipc/registerUserPreferencesIpcHandlers.spec.ts]: Implemented in commit `6a2dbae`.

I added the missing formula-by-input-count IPC coverage in `registerUserPreferencesIpcHandlers.spec.ts`:
- registration assertions for `getCalculatedChannelLastFormulaByInputCount` and `updateCalculatedChannelLastFormulaByInputCount`
- functional tests for both handlers (including payload forwarding for update)

Also re-ran the focused suite:
`npm run test:unit:main-infrastructure -- src/test/main/user-preferences/infrastructure/ipc/registerUserPreferencesIpcHandlers.spec.ts` (pass).

File: src/app/renderer/utils/calculatedChannels.ts [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+      }
+
+      const tokenValue = formula.slice(start, index);
+      const parsed = Number(tokenValue);
+
+      if (!Number.isFinite(parsed)) {
+        throw new Error(`Invalid number: ${tokenValue}`);
+      }
+
+      tokens.push({ kind: 'number', value: parsed });
```
[the remark this answers is not in the document]
Comment (jarosat) [reply src/app/renderer/utils/calculatedChannels.ts]: Implemented in commit `6a2dbae`.

I added a digit-presence guard in the tokenizer before `Number(...)` conversion:
- reject numeric tokens without digits (e.g. `.`) with `Invalid number: .`

Also added regression coverage in `calculatedChannels.spec.ts` (`rejects standalone dot as invalid number`) and re-ran focused renderer unit tests (pass).
